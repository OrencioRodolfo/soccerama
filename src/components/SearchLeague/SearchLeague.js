import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Select from 'material-ui/Select'
import { MenuItem } from 'material-ui/Menu'
import { InputLabel } from 'material-ui/Input'
import { FormControl } from 'material-ui/Form'
import actions from '../../store/actions'

const Form = styled.form`
  display: flex;
  flex-direction: row;
  width: 50%;
  margin: 15px 0;
`

class SearchLeague extends Component {
  static propTypes = {
    queryStanding: PropTypes.func,
    getLeagues: PropTypes.func.isRequired,
    leagues: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      current_season_id: PropTypes.number,
      seasons: PropTypes.shape({
        data: PropTypes.arrayOf(PropTypes.shape({
          id: PropTypes.number,
          name: PropTypes.string,
        })),
      }),
    })).isRequired,
  }

  static defaultProps = {
    queryStanding: () => {},
  }

  state = {
    league: -1,
    season: -1,
  }

  componentDidMount() {
    this.props.getLeagues().then((leagues) => {
      if (leagues.value.data.length) {
        const league = leagues.value.data[0]
        this.setState({
          league: league.id,
          season: league.current_season_id,
        })

        this.fetchStandings(league.id, league.current_season_id)
      }
    })
  }

  findLeagueById = id => this.props.leagues.find(item => item.id === id)

  findSeasonById = (seasons, id) => seasons.find(item => item.id === id)

  handleLeagueChange = (event) => {
    const leagueId = event.target.value
    const league = this.findLeagueById(leagueId)
    this.setState({
      league: leagueId,
      season: league.current_season_id,
    })
    this.fetchStandings(league.id, league.current_season_id)
  }

  handleSeasonChange = (event) => {
    this.setState({
      season: event.target.value,
    }, () => {
      const leagueId = this.state.league
      const seasonId = this.state.season
      if (leagueId >= 0 && seasonId >= 0) {
        this.fetchStandings(leagueId, seasonId)
      }
    })
  }

  fetchStandings = (leagueId, seasonId) => {
    const league = this.findLeagueById(leagueId)
    const season = this.findSeasonById(league.seasons.data, seasonId)
    this.props.queryStanding(league, season)
  }

  renderLeagues() {
    return this.props.leagues.map(league => (
      <MenuItem key={league.id} value={league.id}>
        <em>{league.name}</em>
      </MenuItem>
    ))
  }

  renderSeasons() {
    const league = this.findLeagueById(this.state.league)

    return league.seasons.data.map(season => (
      <MenuItem key={season.id} value={season.id}>
        <em>{season.name}</em>
      </MenuItem>
    ))
  }

  render() {
    return (
      <Form autoComplete="off">
        <FormControl style={{ width: '50%' }}>
          <InputLabel htmlFor="league">League</InputLabel>
          <Select
            style={{ width: '100%' }}
            value={this.state.league}
            onChange={this.handleLeagueChange}
            inputProps={{
              name: 'league',
              id: 'league',
            }}
          >
            {this.renderLeagues()}
          </Select>
        </FormControl>

        {
          (this.state.league >= 0) &&
          <FormControl style={{ width: '50%' }}>
            <InputLabel htmlFor="season">Season</InputLabel>
            <Select
              style={{ width: '100%' }}
              value={this.state.season}
              onChange={this.handleSeasonChange}
              inputProps={{
                name: 'season',
                id: 'season',
              }}
            >
              {this.renderSeasons()}
            </Select>
          </FormControl>
        }
      </Form>
    )
  }
}

const mapStateToProps = ({ league }) => ({
  leagues: league.list,
})

export default connect(mapStateToProps, {
  getLeagues: actions.getLeagues,
})(SearchLeague)

export {
  SearchLeague as DSearchLeague,
  mapStateToProps,
}
