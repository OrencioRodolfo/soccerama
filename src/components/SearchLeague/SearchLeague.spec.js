import React from 'react'
import { shallow } from 'enzyme'
import { DSearchLeague, mapStateToProps } from './SearchLeague'

describe('Search league form component', () => {
  let component
  let getLeagues
  let queryStanding
  let leagues

  beforeEach(() => {
    leagues = [{
      id: 1,
      name: 'sample',
      current_season_id: 1,
      seasons: {
        data: [{
          id: 2,
          name: 'season',
        }],
      },
    }]
    getLeagues = () => Promise.resolve({
      value: {
        data: [{
          id: 1,
          name: 'sample',
          current_season_id: 1,
        }],
      },
    })
    queryStanding = jest.fn()
    component = shallow(<DSearchLeague
      getLeagues={getLeagues}
      leagues={leagues}
      queryStanding={queryStanding}
    />)
  })

  it('Must render properly', () => {
    expect(component).toMatchSnapshot()
  })

  describe('changing league and season in the selects', () => {
    let instance
    beforeEach(() => {
      instance = component.instance()
    })

    it('must add the selected league ID to the state and reset the season', () => {
      instance.handleLeagueChange({ target: { value: 1 } })
      expect(instance.state.league).toEqual(1)
      expect(instance.state.season).toEqual(1)
    })

    it('must add the selected season ID to the state and call queryStanding()', () => {
      instance.setState({ league: 1 })
      instance.handleSeasonChange({ target: { value: 2 } })
      expect(instance.state.season).toEqual(2)
      expect(queryStanding).toBeCalled()
    })
  })

  describe('map state to props', () => {
    it('Must return the league and leagues keys', () => {
      const state = {
        league: {
          details: {},
          list: [],
        },
      }
      const result = mapStateToProps(state)
      expect(result).toEqual({
        leagues: [],
      })
    })
  })
})
