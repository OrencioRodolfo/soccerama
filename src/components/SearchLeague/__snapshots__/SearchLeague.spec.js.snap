// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search league form component Must render properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchLeague
    getLeagues={[Function]}
    leagues={
      Array [
        Object {
          "current_season_id": 1,
          "id": 1,
          "name": "sample",
          "seasons": Object {
            "data": Array [
              Object {
                "id": 2,
                "name": "season",
              },
            ],
          },
        },
      ]
    }
    queryStanding={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "current_season_id": 1,
              "id": 1,
              "name": "sample",
              "seasons": Object {
                "data": Array [
                  Object {
                    "id": 2,
                    "name": "season",
                  },
                ],
              },
            },
            undefined,
          ],
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "autoComplete": "off",
      "children": Array [
        <WithStyles(FormControl)
          style={
            Object {
              "width": "50%",
            }
          }
        >
          <WithStyles(InputLabel)
            htmlFor="league"
          >
            League
          </WithStyles(InputLabel)>
          <WithStyles(Select)
            inputProps={
              Object {
                "id": "league",
                "name": "league",
              }
            }
            onChange={[Function]}
            style={
              Object {
                "width": "100%",
              }
            }
            value={-1}
          >
            <WithStyles(MenuItem)
              value={1}
            >
              <em>
                sample
              </em>
            </WithStyles(MenuItem)>
          </WithStyles(Select)>
        </WithStyles(FormControl)>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(InputLabel)
              htmlFor="league"
            >
              League
            </WithStyles(InputLabel)>,
            <WithStyles(Select)
              inputProps={
                Object {
                  "id": "league",
                  "name": "league",
                }
              }
              onChange={[Function]}
              style={
                Object {
                  "width": "100%",
                }
              }
              value={-1}
            >
              <WithStyles(MenuItem)
                value={1}
              >
                <em>
                  sample
                </em>
              </WithStyles(MenuItem)>
            </WithStyles(Select)>,
          ],
          "style": Object {
            "width": "50%",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "League",
              "htmlFor": "league",
            },
            "ref": null,
            "rendered": "League",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <WithStyles(MenuItem)
                  value={1}
                >
                  <em>
                    sample
                  </em>
                </WithStyles(MenuItem)>,
              ],
              "inputProps": Object {
                "id": "league",
                "name": "league",
              },
              "onChange": [Function],
              "style": Object {
                "width": "100%",
              },
              "value": -1,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "class",
                "props": Object {
                  "children": <em>
                    sample
                  </em>,
                  "value": 1,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "sample",
                  },
                  "ref": null,
                  "rendered": "sample",
                  "type": "em",
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      false,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "autoComplete": "off",
        "children": Array [
          <WithStyles(FormControl)
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <WithStyles(InputLabel)
              htmlFor="league"
            >
              League
            </WithStyles(InputLabel)>
            <WithStyles(Select)
              inputProps={
                Object {
                  "id": "league",
                  "name": "league",
                }
              }
              onChange={[Function]}
              style={
                Object {
                  "width": "100%",
                }
              }
              value={-1}
            >
              <WithStyles(MenuItem)
                value={1}
              >
                <em>
                  sample
                </em>
              </WithStyles(MenuItem)>
            </WithStyles(Select)>
          </WithStyles(FormControl)>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <WithStyles(InputLabel)
                htmlFor="league"
              >
                League
              </WithStyles(InputLabel)>,
              <WithStyles(Select)
                inputProps={
                  Object {
                    "id": "league",
                    "name": "league",
                  }
                }
                onChange={[Function]}
                style={
                  Object {
                    "width": "100%",
                  }
                }
                value={-1}
              >
                <WithStyles(MenuItem)
                  value={1}
                >
                  <em>
                    sample
                  </em>
                </WithStyles(MenuItem)>
              </WithStyles(Select)>,
            ],
            "style": Object {
              "width": "50%",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "League",
                "htmlFor": "league",
              },
              "ref": null,
              "rendered": "League",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <WithStyles(MenuItem)
                    value={1}
                  >
                    <em>
                      sample
                    </em>
                  </WithStyles(MenuItem)>,
                ],
                "inputProps": Object {
                  "id": "league",
                  "name": "league",
                },
                "onChange": [Function],
                "style": Object {
                  "width": "100%",
                },
                "value": -1,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "class",
                  "props": Object {
                    "children": <em>
                      sample
                    </em>,
                    "value": 1,
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "sample",
                    },
                    "ref": null,
                    "rendered": "sample",
                    "type": "em",
                  },
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        false,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
